src/exchange/order_handler.cairo

```diff
@@ -177,50 +214,202 @@ mod OrderHandler {
+            order_event_utils::emit_order_updated(
+                base_order_handler_state.event_emitter.read(),
+                key,
+                size_delta_usd,
+                acceptable_price,
+                trigger_price,
+                min_output_amount,
+            );
+
+            global_reentrancy_guard::non_reentrant_after(data_store);
+
+            order
         }
 
-        fn cancel_order(ref self: ContractState, key: felt252) { // TODO
+        fn cancel_order(ref self: ContractState, key: felt252) {
+            // Check only controller.
+            let role_module_state = RoleModule::unsafe_new_contract_state();
+            role_module_state.only_controller();
+
+            // Fetch data store.
+            let base_order_handler_state = BaseOrderHandler::unsafe_new_contract_state();
+            let data_store = base_order_handler_state.data_store.read();
+
+            global_reentrancy_guard::non_reentrant_before(data_store);
+
+            let order = order_store_utils::get(data_store, key);
+
+            // Validate feature.
+            feature_utils::validate_feature(
+                data_store,
+                keys::cancel_order_feature_disabled_key(get_contract_address(), order.order_type)
+            );
+
+            if base_order_utils::is_market_order(order.order_type) {
+                exchange_utils::validate_request_cancellation(
+                    data_store, order.updated_at_block, 'Order'
+                )
+            }
+
+            order_utils::cancel_order(
+                data_store,
+                base_order_handler_state.event_emitter.read(),
+                base_order_handler_state.order_vault.read(),
+                key,
+                order.account,
+                // starting_gas,  // DISABLED
+                keys::user_initiated_cancel(),
+            );
+
+            global_reentrancy_guard::non_reentrant_after(data_store);
         }
 
-        fn execute_order(
-            ref self: ContractState, key: felt252, oracle_params: SetPricesParams
-        ) { // TODO
+        fn execute_order(ref self: ContractState, key: felt252, oracle_params: SetPricesParams) {
+            // Check only order keeper.
+            let role_module_state = RoleModule::unsafe_new_contract_state();
+            role_module_state.only_order_keeper();
+
+            // Fetch data store.
+            let base_order_handler_state = BaseOrderHandler::unsafe_new_contract_state();
+            let data_store = base_order_handler_state.data_store.read();
+
+            global_reentrancy_guard::non_reentrant_before(data_store);
+            oracle_modules::with_oracle_prices_before(
+                base_order_handler_state.oracle.read(),
+                data_store,
+                base_order_handler_state.event_emitter.read(),
+                oracle_params.clone()
+            );
+
+            // TODO: Did not implement starting gas and try / catch logic as not available in Cairo
```

Comment:
Good thing that you put a TODO here, we'll need to find a solution for that later